### PR TITLE
fix minute parsing

### LINKED ISSUE #95: ValueError: Format specifier missing precision
Hi,

In line 86 and 91 of the examples/gps_simpletest.py, I am receiving the error "ValueError: Format specifier missing precision".

Could this be resolved please?

Thanks!

### PR DESCRIPTION
**Pull request by jkittner**
This fixes the issue that came up in #99. resolves #95 

I tried to simplify this a little to get the best possible precision. We cannot be more accurate than what's in the actual NMEA message.
The previous approach was to parse the sentence (which contains degrees and minutes) to decimal degrees and then convert them back. During this conversion they forgot to divide by 60, which caused the incorrect minutes (as seen in #99).
Since we already get the degrees and minutes, we can directly parse them and provide them without any loss of precision.

I added tests, that verify, that for all sentence types, the degrees and minutes match exactly the values provided in the original raw NMEA message.
- this is the chip's docs: https://drive.google.com/file/d/1qRh-sf5UiobH6Udk-N1_fyKjUmWYsam8/view
- some other docs about the senteces: https://www.rfwireless-world.com/Terminology/GPS-sentences-or-NMEA-sentences.html
- more docs: http://aprs.gids.nl/nmea/#gga

They do not almost match, when it comes to the number of digits, but that's maybe due to different versions?

### CODE REVIEW

File: tests/adafruit_gps_test.py [hunk 1]
```diff
@@ -177,6 +177,10 @@ def test_GPS_update_rmc_no_magnetic_variation():
         assert gps.timestamp_utc == exp_time
         assert gps.latitude == pytest.approx(12.57613)
         assert gps.longitude == pytest.approx(1.385391)
+        assert gps.latitude_degrees == 12
+        assert gps.longitude_degrees == 1
+        assert gps.latitude_minutes == 34.5678
+        assert gps.longitude_minutes == 3.12345
```
Comment (tekktrik) [inline tests/adafruit_gps_test.py:183]: I believe this should be 23.12346.  I'm surprised it passed, if so.

File: examples/gps_simpletest.py [hunk 2]
```diff
@@ -84,13 +85,13 @@
...
-                gps.latitude_degrees, gps.latitude_minutes
+            "Precise Latitude: {:3.0f} degs, {:2.4f} mins".format(
+                math.floor(gps.latitude_degrees), gps.latitude_minutes
             )
         )
         print(
-            "Precise Longitude: {:2.}{:2.4f} degrees".format(
-                gps.longitude_degrees, gps.longitude_minutes
+            "Precise Longitude: {:3.0f} degs, {:2.4f} mins".format(
+                math.floor(gps.longitude_degrees), gps.longitude_minutes
```
Comment (tekktrik) [inline examples/gps_simpletest.py:94]: I don't think you need the float specifier for the degrees if you're using `math.floor()` right?

**Review by tekktrik — CHANGES_REQUESTED** [review]
This looks good to me, would definitely be good to test out on the hardware too, but a few questions in the meantime below.

File: tests/adafruit_gps_test.py (hunk 1 above)
Comment (jkittner) [reply tests/adafruit_gps_test.py:183]: Good catch. That's a bug when parsing the NMEA sentence. I'll try to fix it.

**Comment by dhalbert** [comment]
Hi @jkittner Do you have time to finish this off? Thanks.

**Comment by jkittner** [comment]
@tekktrik @dhalbert - I came up with a temporary fix for this. Things should now work as expected.

**Comment by jkittner** [comment]
@tekktrik @dhalbert do you maybe have time to have a final look at this?

File: tests/adafruit_gps_test.py [hunk 3]
```diff
@@ -467,3 +484,19 @@ def test_GPS_update_from_GSV_both_parts_sats_are_removed():
             assert gps.update()
             assert gps.satellites == 2
             assert set(gps.sats.keys()) == {"GP12", "GP14", "GP13", "GP15"}
+
+
+@pytest.mark.parametrize(
+    ("input_str", "exp", "neg"),
+    (
+        (["3723.2475", "N"], (37, 23.2475), "S"),
```
Comment (tekktrik) [inline tests/adafruit_gps_test.py:492]: The `neg` values should be lower case, I believe, based on the comparison in `_read_deg_mins()`.

**Review by tekktrik — CHANGES_REQUESTED** [review]
Sorry for the delay!  Feedback added but this looks great!  Just a few changes and questions!

File: .pre-commit-config.yaml [hunk 4]
```diff
@@ -4,21 +4,21 @@
 
 repos:
   - repo: https://github.com/python/black
-    rev: 23.3.0
+    rev: 23.10.1
```
[the remark this answers is not in the document]
Comment (jkittner) [reply .pre-commit-config.yaml:7]: alright I'm fine to revert this - no worries - is the autoupdate handled somewhere or how are things kept up to date?

File: adafruit_gps.py [hunk 5]
```diff
@@ -490,26 +500,30 @@ def _parse_gll(self, data: List[str]) -> bool:
...
+        parsed_data = _parse_data(_GLL, data)
         if data is None:
             return False  # Params didn't parse
 
         # Latitude
-        self.latitude = _read_degrees(data, 0, "s")
-        self.latitude_degrees, self.latitude_minutes = _read_int_degrees(data, 0, "s")
+        self.latitude = _read_degrees(parsed_data, 0, "s")
+        self.latitude_degrees, self.latitude_minutes = _read_deg_mins(
+            data=data, index=0, neg="s"
```
[the remark this answers is not in the document]
Comment (jkittner) [reply adafruit_gps.py:510]: yeah, `_read_deg_mins` needs the sting version, while `_read_degrees` needs the integer version (which is returned by `_parse_data`) - I am not too happy with this data structure - this should ideally be a tuple or even better a `NamedTuple` - but the mutability is unfortunately needed afaict.

File: adafruit_gps.py [hunk 6]
```diff
@@ -125,12 +125,22 @@ def _read_degrees(data: List[float], index: int, neg: str) -> float:
...
 
-def _read_int_degrees(data: List[float], index: int, neg: str) -> Tuple[int, float]:
-    deg = data[index] // 1000000
-    minutes = data[index] % 1000000 / 10000
+def _read_deg_mins(data: List[str], index: int, neg: str) -> Tuple[int, float]:
+    # the degrees come in different formats and vary between latitudes and
+    # longitudes, which makes parsing tricky:
+    # for latitudes: ddmm,mmmm (0 - 7 decimal places, not zero padded)
+    # for longitudes: dddmm,mmmm (0 - 7 decimal places, not zero padded)
+    int_part, _, minutes_decimal = data[index].partition(".")
```
[the remark this answers is not in the document]
Comment (jkittner) [reply adafruit_gps.py:133]: `partition` will always return `tuple[str, str, str] so it's save to unpack. `split` will return `list[str]` so we cannot safely unpack - this was my though behind this after working a lot with `mypy`...

File: tox.ini [hunk 7]
```diff
@@ -5,7 +5,7 @@
 envlist = py37
 
 [testenv]
-deps = -rrequirements-dev.txt
+deps = -roptional_requirements.txt
```
[the remark this answers is not in the document]
Comment (jkittner) [reply tox.ini:8]: yeah, this was originally for testing locally. This was before tests were handled somewhere else. If you agree we can simply remove `tox`.
On a more generic note: does the centralized gha handle test coverage though?

File: tests/adafruit_gps_test.py (hunk 3 above)
Comment (jkittner) [reply tests/adafruit_gps_test.py:492]: yeah, we use `.lower` inside `_read_deg_mins` so both will work. I think the NMEA-Standard has them upper case all the time. I'd suggest leaving it as is?

File: adafruit_gps.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (tekktrik) [reply adafruit_gps.py:133]: The docs say that `partition()` isn't available on small boards, though I'm not sure if that's up to date.  If that's the case and this library can run on smaller boards, it may be worth using `split()` if there's ever only one period in the string, which seems to be the case.

File: tests/adafruit_gps_test.py (hunk 3 above)
Comment (tekktrik) [reply tests/adafruit_gps_test.py:492]: I believe the provided `neg` parameter needs to already be lower case while the letter in the string is made lower case for the comparison.  Otherwise it'll compare a coerced lower case in the string to a provided upper case.

File: .pre-commit-config.yaml (hunk 4 above)
[the remark this answers is not in the document]
Comment (tekktrik) [reply .pre-commit-config.yaml]: We typically use adabot to handle library-wide patches

File: adafruit_gps.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (tekktrik) [reply adafruit_gps.py:133]: Yeah, since it's disabled for small boards, let's use `split()`

[the remark this answers is not in the document]
Comment (jkittner) [reply adafruit_gps.py]: mhm, that's a bit annoying but nothing we can do I guess - I found another edge case where partition is needed. It's when the value is provided as an integer - this test-case was testing it `(["00123", "e"], (1, 23), "w")`. That only works with partition.
I had to add some extra code to make this work with `split`

File: tests/adafruit_gps_test.py (hunk 3 above)
Comment (jkittner) [reply tests/adafruit_gps_test.py]: okay - changed.

**Comment by jkittner** [comment]
The `pylint` version is so old it doesn't work with my python version (3.12) anymore so I had to use `-- no-verify` and of course I violated the black style - should have used `SKIP=pylint` instaed...

File: tox.ini (hunk 7 above)
[the remark this answers is not in the document]
Comment (tekktrik) [reply tox.ini]: I don't believe it does coverage, but that would be awesome if it did!  Worth considering for sure.

**Review by tekktrik — APPROVED** [review]
